### DISCUSSION TITLE
getServerSideProps props empty object

### DISCUSSION THREAD

**Discussion by ChristopherLMiller** [vercel/next.js#11873] [Help]
Working on trying to use getServerSideProps in a page [slug].tsx and in the function, the data is defined and can console log it and see the data, however in the actual page if I try and output that data its always undefined and props seems to be just {}.  I'm not sure if my syntax is wrong here or if there is something deeper going on.

```jsx
import ReactMarkdown from 'react-markdown';
import Link from 'next/link';
import Card from '../../components/Card';
import { FullArticle } from '../../components/articles/Full';
import { NextPage, GetServerSideProps } from 'next';
import { Layout } from '../../components/layout/PageLayout';

// @ts-ignore
const PostPage: NextPage = ({data}) => {
  console.log(data); // outputs undefined

  if (data !== undefined) {
    if (data.articles.length >= 1) {
      return (
        <Layout meta={{title, description, useSEO: false}}>
          // @ts-ignore
          {data !== undefined && data.articles.map(article => (
              <FullArticle article={article}>
                <ReactMarkdown source={article.content} escapeHtml={false} />
              </FullArticle>
            ))}
        </Layout>
      );
    }
  }
};

export const getServerSideProps: GetServerSideProps = async context =>  {
  const { slug } = context.query;

  const response = await fetch(url);
  const data = await response.json();
  return { props: { data }}

}

export default PostPage;
```

Is there something obvious I'm missing?

**Comment by ankri** [comment]
Do you have a custom `_app.tsx` ?

Can you please check, that you pass the `pageProps` to your custom `Component`?

```tsx
// pages/_app.tsx
export default ({ Component, pageProps }) => (
  <Component {...pageProps} />
)
```

**Reply by JoelVerastegui** [reply]
You solve my day, thanks ;3

**Reply by KishanPatelGit** [reply]
This is working for me 

`const PostPage: NextPage = ({data}) => {
const router = useRouter();
  if (router.isFallback) {
    return <h3>loading...</h3>;
  }
console.log(data); //It print your data from prop
}`

**Reply by jsparkdev** [reply]
thanks

**Comment by ChristopherLMiller** [comment]
I am indeed.  Although maybe the _app.tsx needs an overhaul from when app was first built as i'm seeing other things that don't seem to be the standard way of handling them anymore.

**Reply by leerob** [reply]
This discussion might be related: https://github.com/zeit/next.js/discussions/12368.

**Comment by ceyhunkeklik** [comment]
I faced same issue on **next@9.3.5**. I think it is a bug actually.

I fixed this issue manually;
If you are using **_app.js** file with the getInitialProps property, probably you are calling "**getInitialProps**" from **NextApp (next/app)** like the following.

```
const appProps = await NextApp.getInitialProps(appContext);

return {
   ...appProps,
}
```

The issue in here. Because, **pageProps** object is empty object and probably it override the original one.

I've updated these lines like the following;

```
const { pageProps, ...appProps } = await NextApp.getInitialProps(appContext);

const result = {
   ...appProps,
}

if (Object.keys(pageProps) > 0) {
    result.pageProps = pageProps;
}

return result,
```

This works for me. 🎉

**Comment by RehaOk** [comment]
The code below works by the way.

` 
static async getInitialProps({Component, ctx}) {

        let pageProps = Component.getInitialProps ? await Component.getInitialProps(ctx) : {};
        if (Object.keys(pageProps).length > 0) {
            return {pageProps: pageProps}; // 1. IF YOU RETURN HERE LIKE THIS(named) WHEN PAGE 
                                        // PROPS IS EMPTY OBJECT aka length === 0
        } else {
            return {};
        } 
    }

    render() {
        const {Component, pageProps, store} = this.props; // 2. this.props will return empty object for pageProps

        return (
            <Provider store={store}>
                <Component {...pageProps}/>
            </Provider>
        );
    }`
Instead **"return {}"** directly when you don't have getInitialProps on the page or when you don't have any page props this way pageProps will get content from getStaticProps.

**Reply by RehaOk** [reply]
This is my first git comment so i didn't quite understand why git doesn't accept 
static async getInitialProps({Component, ctx}) { in the code block sorry for that.

**Reply by geryit** [reply]
thanks, this was the correct way @RehaOk

**Reply by wasurocks** [reply]
This works can confirm. Thank you very much!

**Reply by NNicoletti93** [reply]
This worked for me! Thank you very much!

**Reply by thisames** [reply]
This works can confirm. Thank you very much!!!

**Comment by remjx** [comment]
Make sure you are not using `getServerSideProps` inside a component, it can only be used in the top-level page in `/pages`

**Comment by codersgyan** [comment]
Well, Make sure that you don’t have any non page component in pages directory ✅

**Comment by DusanDjordjic** [comment]
I had the same problem and I realized that I was using` getServerSideProps` in the component that is not the actual page, rather it was my card component that I was displaying on the page. So maybe you made the same silly mistake.

**Reply by ninsau** [reply]
thank you ! this was the same problem I had

**Comment by softmarshmallow** [comment]
I'm having this issue only on production build.

`yarn dev` - works
`yarn build & yarn start` - suddenly does not work


## Opps!
It was the .env problem check sure if you're not using any env vars in getServerSideProps and returning values conditionally.
